phd-tests/tests/src/stats.rs

```diff
@@ -0,0 +1,501 @@
+                    amount,
+                    last_sample.measurement.timestamp()
+                );
+                vcpu_metrics.metrics.insert(state, amount);
+                self.update_metric_times(last_sample.measurement.timestamp());
+            }
+        }
+    }
+}
+
+impl FakeNexusContext {
+    fn new() -> Self {
+        Self { sampler: Arc::new(Mutex::new(None)) }
+    }
+
+    fn set_producer_info(&self, info: ProducerEndpoint) {
+        assert_eq!(info.kind, ProducerKind::Instance);
+        *self.sampler.lock().unwrap() =
+            Some(PropolisOximeterSampler { addr: info.address, uuid: info.id });
+    }
+
+    async fn wait_for_producer(&self) {
+        loop {
+            {
+                let sampler = self.sampler.lock().unwrap();
+                if sampler.is_some() {
+                    return;
+                }
+            }
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
+    }
+
+    /// Sample Propolis' Oximeter metrics, waiting up to a few seconds so that
+    /// all measurements are from the time this function was called or later.
+    async fn wait_for_propolis_stats(&self) -> VirtualMachineMetrics {
+        let min_metric_time = Utc::now();
+
+        let result = backoff::future::retry(
+            backoff::ExponentialBackoff {
+                max_interval: Duration::from_secs(1),
+                max_elapsed_time: Some(Duration::from_secs(10)),
+                ..Default::default()
+            },
+            || async {
+                if let Some(metrics) = self.sample_propolis_stats().await {
+                    if metrics.oldest_time >= min_metric_time {
+                        Ok(metrics)
+                    } else {
+                        Err(backoff::Error::transient(anyhow::anyhow!(
+                            "sampled metrics are not recent enough"
+                        )))
+                    }
+                } else {
+                    Err(backoff::Error::transient(anyhow::anyhow!(
+                        "full metrics sample not available (yet?)"
+                    )))
+                }
+            },
+        )
+        .await;
+
+        result.expect("propolis-server Oximeter stats should become available")
+    }
+
+    /// Sample Propolis' Oximeter metrics, including the timestamp of the oldest
+    /// metric reflected in the sample.
+    ///
+    /// Returns `None` for some kinds of incomplete stats or when no stats are
+    /// available at all.
+    async fn sample_propolis_stats(&self) -> Option<VirtualMachineMetrics> {
+        let metrics_url = {
+            let sampler = self.sampler.lock().unwrap();
+            let stats = sampler.as_ref().expect("stats url info exists");
+            format!("http://{}/{}", stats.addr, stats.uuid)
+        };
+        let res = reqwest::Client::new()
+            .get(metrics_url)
+            .send()
+            .await
+            .expect("can send oximeter stats request");
```

Comment:
hey @bnaecker how much of a crime _is_ this, in your Oximeter eyes? to me it ffeels kinda bad to embed how to query the Oximeter producer here in PHD, but at least this ought to break loudly enough if it stopped returning the expected stats.

Eliza was also talking about having something like an Oximeter test crate [here](https://github.com/oxidecomputer/propolis/pull/855#discussion_r1951654030), and having to make the metrics URL here is really what sells me on the idea, but maybe we expect this to be more stable than i'm giving it credit for.

Reply:
This URL is very stable, but I can see why one would neither know that nor want to rely on it. The producer server one stands up to emit metrics fixes the URL for you, there's zero client control over it today. But as I mentioned on the other thread, the only reason we don't already have the ability for you to extract the samples that oximeter collects already is because we haven't needed it yet. Printing or inserting into the DB has been sufficient.

I think adding functionality to give `oximeter` your own sink or channel or whatever, where it will forward any samples it collects, seems like a fine and and valuable tool to add.